tests/lib/prepare-uc20.sh

```diff
@@ -257,26 +255,49 @@ echo "ubuntu:x:1001:" >> /root/test-var/lib/extrausers/group
 sed -r -i -e 's/^systemd-journal:x:([0-9]+):$/systemd-journal:x:\1:test/' /root/test-etc/group
 
 # mount fresh image and add all our SPREAD_PROJECT data
-kpartx -avs pc.img
-# losetup --list --noheadings returns:
-# /dev/loop1   0 0  1  1 /var/lib/snapd/snaps/ohmygiraffe_3.snap                0     512
-# /dev/loop57  0 0  1  1 /var/lib/snapd/snaps/http_25.snap                      0     512
-# /dev/loop19  0 0  1  1 /var/lib/snapd/snaps/test-snapd-netplan-apply_75.snap  0     512
-devloop=$(losetup --list --noheadings | grep pc.img | awk '{print $1}')
-dev=$(basename "$devloop")
-
-# mount it so we can use it now
-mkdir -p /mnt
-mount "/dev/mapper/${dev}p2" /mnt
-
-# add the data that snapd.spread-tests-run-mode-tweaks.service reads to the 
-# mounted partition
-tar -c -z \
-    -f /mnt/run-mode-overlay-data.tar.gz \
-    /root/test-etc /root/test-var/lib/extrausers
-
-# tear down the mounts
-umount /mnt
-kpartx -d pc.img
+# for the lxd backend this step is a bit different as the device mapper kernel module
+# is not supported by lxd containers. We thus have to do some manual setup of the image
+# partition mount.
+if [ "${SPREAD_BACKEND}" = "lxd-nested" ]; then
+    # Manually try to find to EFI system partition and it's size. The assumption made here
+    # is that the EFI system partition is the first one in the disk. So we mount the partition
+    # that comes right after. We do this as we can't partprobe without it failing on creating
+    # new loop devices for partitions.
+    partoffset=$(fdisk -lu pc.img | awk '/EFI System$/ {print $2}')
+    devloop=$(losetup --show -f pc.img -o $(($partoffset * 512)))
+    mkdir /mnt/p2
+    mount $devloop /mnt/p2
+
+    # add the data that snapd.spread-tests-run-mode-tweaks.service reads to the 
+    # mounted partition
+    tar -c -z \
+        -f /mnt/p2/run-mode-overlay-data.tar.gz \
+        /root/test-etc /root/test-var/lib/extrausers
+
+    umount /mnt/p2
+    losetup -d $devloop
+else
+    kpartx -avs pc.img
+    # losetup --list --noheadings returns:
+    # /dev/loop1   0 0  1  1 /var/lib/snapd/snaps/ohmygiraffe_3.snap                0     512
+    # /dev/loop57  0 0  1  1 /var/lib/snapd/snaps/http_25.snap                      0     512
+    # /dev/loop19  0 0  1  1 /var/lib/snapd/snaps/test-snapd-netplan-apply_75.snap  0     512
+    devloop=$(losetup --list --noheadings | grep pc.img | awk '{print $1}')
+    dev=$(basename "$devloop")
+
+    # mount it so we can use it now
+    mkdir -p /mnt
+    mount "/dev/mapper/${dev}p2" /mnt
+
+    # add the data that snapd.spread-tests-run-mode-tweaks.service reads to the 
+    # mounted partition
+    tar -c -z \
+        -f /mnt/run-mode-overlay-data.tar.gz \
+        /root/test-etc /root/test-var/lib/extrausers
+
+    # tear down the mounts
+    umount /mnt
+    kpartx -d pc.img
```

Comment:
It would be nice to do this also in a term trap to make sure the unmounting happens also if something fails between mounting and unmounting:

    f() {
        umount...
    }
    trap f TERM

But this is just a suggestion, if this happens in a temporary container then it does not matter too much.

Reply:
or `trap f EXIT` so that it works not only for `SIGTERM`.

Reply:
It is executed either in the LXD container or a VM - so it should not cause any issues, however I still believe we should apply this as it would be more proper. Thanks for your suggestions alfonso, greatly appreciated